crates/circuit/src/operations.rs

```diff
@@ -0,0 +1,848 @@
+            Self::Operation(op) => op.standard_gate(),
+        }
+    }
+
+    fn directive(&self) -> bool {
+        match self {
+            Self::Standard(op) => op.directive(),
+            Self::Gate(op) => op.directive(),
+            Self::Instruction(op) => op.directive(),
+            Self::Operation(op) => op.directive(),
+        }
+    }
+}
+
+/// Trait for generic circuit operations these define the common attributes
+/// needed for something to be addable to the circuit struct
+pub trait Operation {
+    fn name(&self) -> &str;
+    fn num_qubits(&self) -> u32;
+    fn num_clbits(&self) -> u32;
+    fn num_params(&self) -> u32;
+    fn control_flow(&self) -> bool;
+    fn matrix(&self, params: Option<SmallVec<[Param; 3]>>) -> Option<Array2<Complex64>>;
+    fn definition(&self, params: Option<SmallVec<[Param; 3]>>) -> Option<CircuitData>;
+    fn standard_gate(&self) -> Option<StandardGate>;
+    fn directive(&self) -> bool;
+}
+
+#[derive(Clone, Debug)]
+pub enum Param {
+    ParameterExpression(PyObject),
+    Float(f64),
+    Obj(PyObject),
+}
+
+impl<'py> FromPyObject<'py> for Param {
+    fn extract_bound(b: &Bound<'py, PyAny>) -> Result<Self, PyErr> {
+        let param_class = PARAMETER_EXPRESSION
+            .get_or_init(b.py(), || {
+                PyModule::import_bound(b.py(), "qiskit.circuit.parameterexpression")
+                    .unwrap()
+                    .getattr("ParameterExpression")
+                    .unwrap()
+                    .unbind()
+            })
+            .bind(b.py());
+        let circuit_class = QUANTUM_CIRCUIT
+            .get_or_init(b.py(), || {
+                PyModule::import_bound(b.py(), "qiskit.circuit.quantumcircuit")
+                    .unwrap()
+                    .getattr("QuantumCircuit")
+                    .unwrap()
+                    .unbind()
+            })
+            .bind(b.py());
+        Ok(
+            if b.is_instance(param_class)? || b.is_instance(circuit_class)? {
+                Param::ParameterExpression(b.clone().unbind())
+            } else if let Ok(val) = b.extract::<f64>() {
+                Param::Float(val)
+            } else {
+                Param::Obj(b.clone().unbind())
+            },
+        )
+    }
+}
+
+impl IntoPy<PyObject> for Param {
+    fn into_py(self, py: Python) -> PyObject {
+        match &self {
+            Self::Float(val) => val.to_object(py),
+            Self::ParameterExpression(val) => val.clone_ref(py),
+            Self::Obj(val) => val.clone_ref(py),
+        }
+    }
+}
+
+impl ToPyObject for Param {
+    fn to_object(&self, py: Python) -> PyObject {
+        match self {
+            Self::Float(val) => val.to_object(py),
+            Self::ParameterExpression(val) => val.clone_ref(py),
+            Self::Obj(val) => val.clone_ref(py),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Copy, Eq, PartialEq, Hash)]
+#[pyclass(module = "qiskit._accelerate.circuit")]
+pub enum StandardGate {
```

Comment:
This is a good question. It would be useful if a user could implement an efficient gate that gets the same optimized code paths that `StandardGate` might have. Although at the moment there is not much optimized code for `StandardGate`. But `enum` is not extensible in Rust.

Reply:
I did this because this is the most efficient way I could think of to represent a gate. In practice Qiskit's standard gate library is relatively fixed we don't add gates very frequently and the other circuit objects are all typical built from them. In the transpiler as well we typically are normalizing the circuit to something in the standard gate library very early in the pipeline. So making sure the data structure is efficient here was a high priority. The list is incomplete because I picked a small subset (semi-randomly) to avoid making this PR gigantic (it is already ~4k lines with an incomplete set of gates). After this merges I plan to open good first rust issues that target adding the missing gates.

While this is not the most ergonomic design pattern for extending we don't add standard gates very often. For a extensible pattern this is what the `Operation` trait is for as it gives us an interface definition for what a gate looks like and if we want to add future non-standard gates defined in rust we can use that.

Reply:
I understand that was a relatively quick way to move forward, and already has been a very large undertaking. I already added one gate the `RGate`. It was kind of a pain because the implementation is not very ergonomic, although I don't see an obvious way to make it more so (well there are a few things, like the initialization of the arrays of 19 `None`s). Getting people to fill in the standard gates will involve boiler plate, but is doable.

I see that there are provisions for arbitrary gates via `Operation`. But that's not what I am interested in. I am concerned that the set of high-performance gates is arbitrary, not very convenient to extend in the library, and impossible to extend outside of it.

Reply:
This wasn't about speed of implementation, it's actually a bit more annoying to work with than some other possible representations. What I was talking about is efficiency in implementation. This path ends up with each standard gate being represented by a single byte  (unless we go over 255 gates; which is unlikely because we're only at 40 or so currently I think) and all the attributes are all statically defined. This will likely have the lowest memory overhead and be the fastest to execute in practice. It was about optimizing the code for the common case so that what most people use are as efficient as we can be. The tradeoff is as you say it's not the most ergonomic and if we need to extend it we have to do that explicitly. But we very rarely add standard gates to the library so I was fine with this tradeoff, it's only the initial implementation that's a pain.

Adding extensibility from Rust here would be by doing something like adding `Box<dyn Operation>` as an allowed type  to `OperationType`: the type for `CircuitInstruction`'s `operation` field. So you can make an efficient representation of anything that implements the trait. This does add overhead because of a layer of indirection for the `Box`, but that's about the best you can do in rust for something like this where we want to allow heterogeneous dynamic types on the same field.